src/models/sourcecodemodel.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+    SPDX-FileCopyrightText: Lieven Hey <lieven.hey@kdab.com>
+    SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+#include "sourcecodemodel.h"
+#include <QDir>
+#include <QFile>
+
+SourceCodeModel::SourceCodeModel(QObject* parent)
+    : QAbstractTableModel(parent)
+{
+}
+
+SourceCodeModel::~SourceCodeModel() = default;
+
+void SourceCodeModel::clear()
+{
+    beginResetModel();
+    m_sourceCode.clear();
+    endResetModel();
+}
+
+void SourceCodeModel::setDisassembly(const DisassemblyOutput& disassemblyOutput)
+{
+    if (disassemblyOutput.sourceFileName.isEmpty())
+        return;
+
+    QFile file(m_sysroot + QDir::separator() + disassemblyOutput.sourceFileName);
+
+    if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
+        return;
+
+    beginResetModel();
+
+    const auto lines = QString::fromUtf8(file.readAll()).split(QLatin1Char('\n'));
+
+    int maxIndex = 0;
+    int minIndex = INT_MAX;
+
+    m_validLineNumbers.clear();
+
+    for (const auto& line : disassemblyOutput.disassemblyLines) {
+        if (line.sourceCodeLine == 0)
+            continue;
+
+        if (line.sourceCodeLine > maxIndex) {
+            maxIndex = line.sourceCodeLine;
+        }
+        if (line.sourceCodeLine < minIndex) {
+            minIndex = line.sourceCodeLine;
+        }
+
+        m_validLineNumbers.insert(line.sourceCodeLine);
+    }
+
+    m_sourceCode.clear();
+
+    // the start / end of a function is defined by { }
+    // the following code tries to find ( to include the function signature in the source code
+    if (!lines[minIndex - 1].contains(QLatin1Char('('))) {
```

Comment:
if minIndex == 0 then this underflows

Reply:
I have to rename that variable. Its not really an index, its the line number. Therefore it can't be 0. I will add some assertions to ensure this